Replace abandoned Sensiolabs security checker

paras-malhotra opened the pull request:
This PR replaces the deprecated [Sensiolabs security checker](https://github.com/sensiolabs/security-checker) with the [Enlightn security checker](https://github.com/enlightn/security-checker). Fixes https://github.com/Jorijn/laravel-security-checker/issues/32.

Jorijn commented on src/Console/SecurityCommand.php:
```diff
@@ -2,9 +2,12 @@
 
 namespace Jorijn\LaravelSecurityChecker\Console;
 
+use Enlightn\SecurityChecker\AdvisoryAnalyzer;
```
In light of DRY: There are now three instances where all four dependencies are used. I'd like for this to move into a separate security service that outputs `$checkResult` like below. I believe this makes it easier to maintain and test since we're maintaining just one class instead of three.

paras-malhotra replied:
That's a good point, I'll make sure to PR this into Enlightn for a better API.

Jorijn commented on src/Console/SecurityCommand.php:
```diff
@@ -43,10 +29,14 @@ public function handle()
 
-        // and feed it into the SecurityChecker
-        $checkResult = json_decode((string)$this->checker->check($composerLock), true);
+        $parser = new AdvisoryParser((new AdvisoryFetcher)->fetchAdvisories());
```
I usually avoid class instantiation in the class itself in favor of Dependency Injection through the constructor. This way, we can mock the dependency in the unit tests.

Jorijn commented on tests/SecurityCommandTest.php:
```diff
@@ -2,46 +2,23 @@
 
-        // and check if it returns exit-code 1 (error state)
-        $this->assertEquals($res, 1);
+        $this->assertEquals(1, $this->artisan(
```
There's a problem with this approach: It actually queries the vulnerability database. I don't like it that an external service can influence whether my tests succeed or not. What happens when a dependency listed as `safe` does get affected somehow? I think it's better to mock the response from `analyzeDependencies` (etc.) so we get a predictable outcome.

paras-malhotra replied:
Sure thing, I'll convert the integration test into a unit test. That would take care of the latter 2 comments. 👍

Jorijn reviewed (changes requested):
Hey! 

Thanks for taking the time to set up this extended PR, really appreciate it. Nice job integrating this into Enlightn. 

I do have some suggestions I'd like to see addressed before merging. I hope you can find the time to do so.

Thanks.

paras-malhotra commented:
@Jorijn all done! I also noticed that the tests only ran on Laravel 5.5 due to versioning issues. I fixed them so that Laravel 8.x is also testable.

Versions from Laravel 5.5 to 8.x are all fine now. This can be verified by a `composer update --prefer-lowest` and a `composer update --prefer-stable`.

Jorijn commented on tests/SecuritySlackTest.php:
```diff
@@ -12,6 +11,10 @@ class SecuritySlackTest extends TestCase
     public function testSlackNotification()
     {
+        if (! class_exists(Illuminate\Notifications\Messages\SlackMessage::class)) {
+            $this->markTestSkipped();
```
```suggestion
            $this->markTestSkipped('this test is skipped because package `laravel/slack-notification-channel` is not installed.');
```

Jorijn commented:
I like the update. Just a thought here: Wouldn't it make sense to add `laravel/slack-notification-channel` to the `require-dev` part of composer.json? You're skipping the test if the `SlackMessage::class` doesn't exist, but in the dev context where PHPUnit always runs, it will always get skipped, rendering the test useless. 

When we add the package to `require-dev`, we can remove the check if the SlackMessage object exists without enforcing an optional package onto the end-user while still consistently checking whether the Slack part of the library still works.

paras-malhotra commented:
@Jorijn the issue is that the `laravel/slack-notification-channel` works on 5.8+. So, if we add the `require-dev` we won't be able to test 5.5. Don't think this will impact installations so I guess we can add it so that the test runs on the latest at least.

Jorijn commented:
@paras-malhotra good point, I haven't been close to Laravel for a while, so this stuff is getting away from me. Looking at Laravel's release cycle, I noticed that 5.5 LTS isn't getting any bug fixes as of August 2020, so I don't see why this package should continue to maintain an EOL version. Composer will probably be able to figure out constraints if people do want to install older versions.

I'm contemplating dropping support for <6.0 altogether; 6.x been LTS since 2019. It seems like you're more involved in Laravel. Is this doable?

paras-malhotra commented:
Yes @Jorijn, I think that's a great idea! The latest versions of most packages are focusing on 6-8x.

Jorijn commented:
@paras-malhotra then it's settled, v2.2 will drop support for Laravel <6.0. Are you able to fix the SlackMessage test and update the composer.json? I can draft a new release later today/early tomorrow.

paras-malhotra commented:
Sure, I'll make the changes and ping you once done. Do you also want to shift to Github actions?

Jorijn commented:
@paras-malhotra Great.

Yes, that's something I intended to do a long time ago. For newer projects, I'm using a combination of GitHub Actions, a SonarQube Quality Gate ([example](https://sonarcloud.io/dashboard?id=Jorijn_bitcoin-dca)) en optionally, Docker.

paras-malhotra commented:
@Jorijn, I've updated the versions and added Github actions. I think you'll need to disable Scrutinizer and enable Github actions on your end. Feel free to close this PR if you think I need to ree

Jorijn commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,34 @@
+        php: [7.3, 7.4, 8.0]
+        laravel: [6.*, 7.*, 8.*]
+
+    name: P${{ matrix.php }} - L${{ matrix.laravel }}
```
```suggestion
    name: PHP ${{ matrix.php }} - Laravel ${{ matrix.laravel }}
```

Being a bit more verbose here.

Jorijn commented:
LGTM. Is this everything?

paras-malhotra commented:
Yes, good to go! 👍

Jorijn commented:
Thanks so much for your help and swift responses!

paras-malhotra commented:
@Jorijn, can you please draft a new release? The Sensiolabs checker has stopped working from Monday and anyone who has this package as a dependency right now is probably getting exception logs.

Jorijn commented:
@paras-malhotra it got away from me, thanks for the ping. I just pushed out v2.2.0.

paras-malhotra commented:
Thanks so much @Jorijn! 🎉
